### PR TITLE
[MLIR][OpenMP] Refactor bounds offsetting and fix to apply to all directives

### PR DESCRIPTION
**Pull request by agozillon**
This PR refactors bounds offsetting by combining the two differing implementations (one applying to initial derived type member map implementation for descriptors and the other for regular arrays, effectively allocatable array vs regular array in fortran) now that it's a little simpler to do.

The PR also moves the utilization of createAlteredByCaptureMap into genMapInfoOp, where it will be correctly applied to all MapInfoData, appropriately offsetting and altering Pointer data set in the kernel argument structure on the host.  This primarily means bounds offsets will now correctly apply to enter/exit/update map clauses as opposed to just the Target directive that is currently the case. A few fortran runtime tests have been added to verify this new behavior.

This PR depends on: https://github.com/llvm/llvm-project/pull/84328 and is an extraction of the larger derived type member map PR stack (so a requirement for it to land).

### CODE REVIEW

**Comment by agozillon** [comment]
Final little extraction into seperate PR from the PR stack of derived type member mapping, it should cut down the noise a little bit. Of note is that this PR is unfortunately dependent on the other for a function call, so it'll fail the CI until the other lands and I can rebase on it. 

I am thinking it may be simpler and less noisy to leave the derived type member map PR stack as is for the time being, until these dependencies land and then rebase out the things that have landed. Which would give me time to perhaps prepare an update for adding the nested member map support. Open to reviewer preferences on this one, I won't be pestering anyone for any reviews on the larger PR stack until these land and it's rebased (although, always welcome).

**Comment by agozillon** [comment]
Updated this PR to be self-sufficient with a more concise implementation of checkIfPointer that works in the current (and hopefully other future) use-cases. 

If this iteration is acceptable and landed it would mean the following PR would be unrequired for the time being and can be closed: https://github.com/llvm/llvm-project/pull/84328

**Comment by agozillon** [comment]
Minor ping for attention on this PR, it's quite useful in the sense that it refactors a bit of the map lowering and allows bounds offsets to apply to enter/exit/update. 

It would also mean I can close the other PR, and update the PR stack for derived type member mapping a little easier (that's the goal at least).

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (jsjodin) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2138]: Can this case be placed first so we don't need the conditional expression above?

**Comment by agozillon** [comment]
Small ping for any other reviewers that are interested in reviewing this PR to make comment (or approve if it looks reasonable external to the currently unaddressed reviewer comments to you)!

I'll aim to address the current comments tomorrow and update the PR to address the conflicts it has with the main branch.

**Comment by agozillon** [comment]
Small lie, I updated the PR much earlier than stated! My internal anxiety to get things out of the way quickly won over unfortunately :-) 

The PR is rebased and the current review comment should be addressed. 

However, I will still happily address further review comments made tomorrow or as soon as I can after the fact, as I would very much like to land this PR so I can rebase the larger PR stack on top of it and simplify at least one part of the stack greatly. Plus this adds some features that are currently missing from enter/exit/update relating to map bounds.

**Review by jsjodin — APPROVED** [review]
LGTM

**Comment by agozillon** [comment]
Thank you @jsjodin! 

I'll look towards landing this tomorrow afternoon (EU time) if there's no further review comments from other reviewers in the meantime.

**Pull request merged**
